### PR TITLE
fix: escaped quotes in regexes do not end the regex

### PR DESCRIPTION
**Pull request by fzipi**
Signed-off-by: Felipe Zipitria <felipe.zipitria@owasp.org>

- Fixes bug when double quotes were escaped in `@rx`
- Convert to json

### CODE REVIEW

**Comment by irvinlim** [comment]
Hey @fzipi! Thanks for the contribution. Unfortunately I no long am working with ModSecurity, and have little remaining knowledge and/or interest in this project from last working on ModSec from almost 5 years ago, haha.

I see you are working at OWASP. Would you like to take over this project?

File: syntaxes/modsecurity.tmLanguage.json [hunk 1]
```diff
@@ -0,0 +1,736 @@
...
+      "captures": {
+        "1": {
+          "name": "keyword.control.modsecurity"
+        },
+        "2": {
+          "name": "string.regexp.modsecurity"
+        }
+      },
+      "comment": "Regexp operator and operand, e.g. \"@rx foo\"",
+      "match": "(?i)\"(@rx)\\s+((?:\\\"|[^\"])*)",
```
Comment (irvinlim) [inline syntaxes/modsecurity.tmLanguage.json]: This RegEx `(?:\"|[^"])` means `"` or non-`"`, which effectively makes it all characters, was that what you were going for?

The regex is tricky for this one.. But this SO answer helps: https://stackoverflow.com/a/249937/2037090

```
(?i)"(@rx)\s+((?:[^"\\]|\\.)*)"
```

Comment (irvinlim) [reply syntaxes/modsecurity.tmLanguage.json]: Looks working when I added an escaped double quote in a CRS rule.

image

Comment (irvinlim) [reply syntaxes/modsecurity.tmLanguage.json]: Btw.. It seems like rules starting with `!@rx` are not handled and are treated as string literals.

image

I really don't remember any ModSecurity syntax any more, but it seems like it should mean "not matching regex" right? I guess we could handle those as well, we could create a separate issue for that.

**Review by irvinlim — CHANGES_REQUESTED** [review]
I think we also have to fix the other rules because double quote escapes will break for string literals, `@pm`, etc.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,9 @@
 # Change Log
 
+## 0.1.3
+
+- Fixed escaped double quotes in `@rx` operator
+
```
Comment (irvinlim) [inline CHANGELOG.md]: Let's also not bump the version and CHANGELOG in the PR, we will do that as a separate commit once changes are finalized

**Comment by fzipi** [comment]
> Hey @fzipi! Thanks for the contribution. Unfortunately I no long am working with ModSecurity, and have little remaining knowledge and/or interest in this project from last working on ModSec from almost 5 years ago, haha.
> 
> I see you are working at OWASP. Would you like to take over this project?

Sure, I can take this project if you want. I'm using it on a daily basis :)

File: syntaxes/modsecurity.tmLanguage.json (hunk 1 above)
Comment (fzipi360) [reply syntaxes/modsecurity.tmLanguage.json:178]: Yes, let's do it in a different PR. I can confirm that this is the behavior, but is not expected.

File: CHANGELOG.md (hunk 2 above)
Comment (fzipi) [reply CHANGELOG.md]: Removed.

File: syntaxes/modsecurity.tmLanguage.json (hunk 1 above)
Comment (fzipi) [reply syntaxes/modsecurity.tmLanguage.json]: Yeah, let's create a new issue and PR with that.

Comment (irvinlim) [reply syntaxes/modsecurity.tmLanguage.json]: Sorry for the confusion. I meant that the regex did not work, you have to use

```
(?i)"(@rx)\s+((?:[^"\\]|\\.)*)"
```

After changing it to this, then it would be as in the screenshot above.

**Comment by irvinlim** [comment]
> Sure, I can take this project if you want. I'm using it on a daily basis :)

Great! I don't mind still doing some maintenance here and there, but I don't foresee myself making and updates or changes simply because I don't use ModSec any more.

I could grant you maintainer permissions to this repository, or if you wanted to, I could even transfer the repository and the VS Code Extension page to you (though I need to find out how to do those)

**Review by irvinlim — CHANGES_REQUESTED** [review]
Pending fix with regex change

File: syntaxes/modsecurity.tmLanguage.json (hunk 1 above)
Comment (fzipi) [reply syntaxes/modsecurity.tmLanguage.json]: Changed. Needed to add escapes for json, but should be equivalent...

Comment (irvinlim) [reply syntaxes/modsecurity.tmLanguage.json]: Hmm, still not right, you need to escape the `\` as well.

After escaping it should be 

```
"match": "(?i)\"(@rx)\\s+((?:[^\"\\\\]|\\\\.)*)\"",
```

Comment (fzipi) [reply syntaxes/modsecurity.tmLanguage.json]: 🤦 Sorry! :) Looks like it is good now.

**Review by irvinlim — APPROVED** [review]

**Comment by irvinlim** [comment]
LGTM

**Comment by irvinlim** [comment]
Thanks @fzipi 😄  let me know which you prefer:

1. I can continue to review and merge PRs, and handle the release
2. Add you as maintainer so you can do so, but I will still do the release (since I think publishing needs my own credentials)
3. Transfer ownership to your name, or perhaps some GitHub org that OWASP manages?

**Comment by fzipi** [comment]
I think this is mostly stable long term. Maybe small fixes in the future, but I would say it will be kind of static after some fixes.

Whatever suits you best. I'm part of the coreruleset organization, and that movement can be done faster than the OWASP movement.

But you can transfer it to me and I can handle the movement to the organization if needed. Honestly, what gives you less headaches..... :)

**Comment by irvinlim** [comment]
Oh I see :) I did a brief search on how to transfer ownership of a VS Code extension, it seems like you can't do it without publishing a whole new extension: https://github.com/microsoft/vscode/issues/21478

Maybe I'll just hold on to this for the time being, that might give the least headaches rather than transferring back and forth 😂  If you and your team have decided whether to take over under the `coreruleset` or `owasp` (or `SpiderLabs`? I'm not sure) GitHub orgs, let me know and I will be cool with that too :) It seems that having "official" OWASP maintainers seems to be the best way forward.
